core/peerset/src/lib.rs

```diff
@@ -103,130 +178,129 @@ pub struct PeersetConfig {
-	/// >			otherwise it will not be able to connect to it.
-	pub fn add_reserved_peer(&self, peer_id: PeerId) {
-		let mut inner = self.inner.lock();
-		if !inner.reserved.insert(peer_id.clone()) {
-			// Immediately return if this peer was already in the list.
-			return;
+		for peer_id in config.bootnodes {
+			peerset.data.discovered.add_peer(peer_id, SlotType::Common);
 		}
 
+		peerset.alloc_slots();
+		(peerset, handle)
+	}
+
+	fn on_add_reserved_peer(&mut self, peer_id: PeerId) {
 		// Nothing more to do if we're already connected.
-		if inner.out_slots.iter().chain(inner.in_slots.iter()).any(|s| s.as_ref() == Some(&peer_id)) {
+		if self.data.in_slots.contains(&peer_id) {
+			self.data.in_slots.mark_reserved(&peer_id);
 			return;
 		}
 
-		// Assign a slot for this reserved peer.
-		if let Some(pos) = inner.out_slots.iter().position(|s| s.as_ref().map(|n| !inner.reserved.contains(n)).unwrap_or(true)) {
-			let _ = self.tx.unbounded_send(Message::Connect(peer_id.clone()));
-			inner.out_slots[pos] = Some(peer_id);
-
-		} else {
-			// All slots are filled with reserved peers.
-			if inner.discovered.iter().all(|p| *p != peer_id) {
-				inner.discovered.push(peer_id);
+		match self.data.out_slots.add_peer(peer_id.clone(), SlotType::Reserved) {
+			Ok(_) => {
+				self.message_queue.push_back(Message::Connect(peer_id));
+			},
+			Err(SlotError::AlreadyConnected(_)) => {
+				return;
+			}
+			Err(SlotError::MaxConnections(_)) => {
+				self.data.discovered.add_peer(peer_id, SlotType::Reserved);
+			}
+			Err(SlotError::DemandReroute { disconnect, ..}) => {
+				self.message_queue.push_back(Message::Drop(disconnect));
+				self.data.discovered.add_peer(peer_id, SlotType::Reserved);
 			}
 		}
 	}
 
-	/// Remove a previously-added reserved peer.
-	///
-	/// Has no effect if the node was not a reserved peer.
-	pub fn remove_reserved_peer(&self, peer_id: &PeerId) {
-		let mut inner = self.inner.lock();
-		inner.reserved.remove(peer_id);
+	fn on_remove_reserved_peer(&mut self, peer_id: &PeerId) {
+		self.data.in_slots.mark_not_reserved(peer_id);
+		self.data.out_slots.mark_not_reserved(peer_id);
+		// TODO: should we disconnect from this peer?
+		// a) always?
+		// b) only if reserved_only is set
 	}
 
-	/// Sets whether or not the peerset only has connections .
-	pub fn set_reserved_only(&self, reserved_only: bool) {
-		let mut inner = self.inner.lock();
-		let inner = &mut *inner;	// Fixes a borrowing issue.
-		inner.reserved_only = reserved_only;
-
+	fn on_set_reserved_only(&mut self, reserved_only: bool) {
 		// Disconnect non-reserved nodes.
-		if reserved_only {
-			for slot in inner.out_slots.iter_mut().chain(inner.in_slots.iter_mut()) {
-				if let Some(peer) = slot.as_ref() {
-					if inner.reserved.contains(peer) {
-						continue;
-					}
-
-					let _ = self.tx.unbounded_send(Message::Drop(peer.clone()));
-				}
-
-				*slot = None;
+		self.data.reserved_only = reserved_only;
+		if self.data.reserved_only {
```

Comment:
I don't think this actually removes the nodes from the slots, does it?

Reply:
It does not, it just creates `Message::Drop(peer_id)` for every node that should be removed. I assume that `peerset.dropped(peer_id)` will be called later on when the peer is actually dropped. Is that assumption correct?

Reply:
> I assume that peerset.dropped(peer_id) will be called later on when the peer is actually dropped. Is that assumption correct?

No! From the point of view of the PSM, connects and disconnects are all instantaneous.